pkg/operator/builder.go

```diff
@@ -0,0 +1,349 @@
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/atlassearchindexconfig"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/atlasstream"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/connectionsecret"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/watch"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/indexer"
+)
+
+const (
+	DefaultAtlasDomain      = "https://cloud.mongodb.com/"
+	DefaultSyncPeriod       = 3 * time.Hour
+	DefaultMetricAddress    = ":8080"
+	DefaultProbeAddress     = ":8081"
+	DefaultLeaderElectionID = "06d035fb.mongodb.com"
+)
+
+type Builder struct {
+	scheme           *runtime.Scheme
+	config           *rest.Config
+	namespaces       []string
+	logger           *zap.Logger
+	syncPeriod       time.Duration
+	metricAddress    string
+	probeAddress     string
+	leaderElection   bool
+	leaderElectionID string
+
+	atlasDomain        string
+	predicates         []predicate.Predicate
+	apiSecret          client.ObjectKey
+	atlasProvider      atlas.Provider
+	featureFlags       *featureflags.FeatureFlags
+	deletionProtection bool
+}
+
+func (b *Builder) WithScheme(scheme *runtime.Scheme) *Builder {
+	b.scheme = scheme
+	return b
+}
+
+func (b *Builder) WithConfig(config *rest.Config) *Builder {
+	b.config = config
+	return b
+}
+
+func (b *Builder) WithNamespaces(namespaces ...string) *Builder {
+	b.namespaces = namespaces
+	return b
+}
+
+func (b *Builder) WithLogger(logger *zap.Logger) *Builder {
+	b.logger = logger
+	return b
+}
+
+func (b *Builder) WithSyncPeriod(period time.Duration) *Builder {
+	b.syncPeriod = period
+	return b
+}
+
+func (b *Builder) WithMetricAddress(address string) *Builder {
+	b.metricAddress = address
+	return b
+}
+
+func (b *Builder) WithProbeAddress(address string) *Builder {
+	b.probeAddress = address
+	return b
+}
+
+func (b *Builder) WithLeaderElection(enable bool) *Builder {
+	b.leaderElection = enable
+	return b
+}
+
+func (b *Builder) WithAtlasDomain(domain string) *Builder {
+	b.atlasDomain = domain
+	return b
+}
+
+func (b *Builder) WithPredicates(predicates []predicate.Predicate) *Builder {
+	b.predicates = predicates
+	return b
+}
+
+func (b *Builder) WithAPISecret(apiSecret client.ObjectKey) *Builder {
+	b.apiSecret = apiSecret
+	return b
+}
+
+func (b *Builder) WithAtlasProvider(provider atlas.Provider) *Builder {
+	b.atlasProvider = provider
+	return b
+}
+
+func (b *Builder) WithFeatureFlags(featureFlags *featureflags.FeatureFlags) *Builder {
+	b.featureFlags = featureFlags
+	return b
+}
+
+func (b *Builder) WithDeletionProtection(deletionProtection bool) *Builder {
+	b.deletionProtection = deletionProtection
+	return b
+}
+
+func (b *Builder) Build(ctx context.Context) (manager.Manager, error) {
+	mergeDefaults(b)
+
+	utilruntime.Must(clientgoscheme.AddToScheme(b.scheme))
+	utilruntime.Must(akov2.AddToScheme(b.scheme))
+
+	cacheOpts := cache.Options{
+		SyncPeriod: &b.syncPeriod,
+	}
+
+	switch len(b.namespaces) {
```

Comment:
i'm a big fan of switch statements but I guess a `if len(b.namespaces) == 0 { ... } else { ... }` is more readable here.